CONTRIBUTING_A_NEW_QUEST.md

```diff
@@ -0,0 +1,497 @@
+
+# Bad documentation is a bug
+
+Unclear documents, including this one, are a bug. Feel free to either submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) with a fix or [open an issue](https://github.com/streetcomplete/StreetComplete/issues/new) describing your confusion.
+
+Note that not everything will be directly described. This document very intentionally doesn't include a step-by-step guide to installing Android Studio, [linking](CONTRIBUTING.md#development) to official docs instead.
+
+# More complexity
+What was described above is an attempt to cover all aspects of quest creation, without describing all the complexity.
+
+Below is some additional info.
+
+## getTitle and getTitleArgs
+
+In some cases it is desirable to change quest title, based on specific object.
+
+This requires preparing space in the message for filling at runtime, and to add the mechanism supplying data to replace placeholders.
+
+### Display value of tag if it is present
+
+[AddFireHydrantDiameter](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/fire_hydrant_diameter/AddFireHydrantDiameter.kt) displays `ref` code of hydrant, if tagged.
+
+```
+    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> =
+        arrayOfNotNull(tags["ref"])
+
+    override fun getTitle(tags: Map<String, String>): Int {
+        val hasRef = tags.containsAnyKey("ref")
+        return when {
+            hasRef -> R.string.quest_fireHydrant_diameter_ref_title
+            else   -> R.string.quest_fireHydrant_diameter_title
+        }
+    }
+```
+
+``getTitleArgs` will return either array with single element containing value of `ref` tag or an empty array. In either cases it will be passed to code displaying title at runtime.
+
+`arrayOfNotNull` is a minor helper function [defined in StreetComplete](app/src/main/java/de/westnordost/streetcomplete/ktx/Collections.kt). Note that when you open code in Android Studio or other IDE you can easily go to place where given function is defined.
+
+``getTitle` provide identifier of text which will be
+
+So `getTitle` function that returns identifier to text in [strings file](app/src/main/res/values/strings.xml). It will be either 
+
+
+* `quest_fireHydrant_diameter_ref_title` which represents `What diameter is specified on the sign for this fire hydrant? (reference number: %s)` (or translation). `%s` is a [placeholder](https://developer.android.com/guide/topics/resources/string-resource.html#formatting-strings) and will be replaced by text provided by `getTitleArgs`
+* In case of ref tag being not present `quest_fireHydrant_diameter_title` (`What diameter is specified on the sign for this fire hydrant?` will be used.
```

Comment:
```suggestion
* In case of ref tag being not present `quest_fireHydrant_diameter_title` (`What diameter is specified on the sign for this fire hydrant?`) will be used.
```